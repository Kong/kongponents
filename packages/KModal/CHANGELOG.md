# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.9](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.7...@kongponents/kmodal@0.1.9) (2020-01-22)

**Note:** Version bump only for package @kongponents/kmodal





## [0.1.6](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-10...@kongponents/kmodal@0.1.6) (2020-01-16)

**Note:** Version bump only for package @kongponents/kmodal





## [0.1.4-10](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-9...@kongponents/kmodal@0.1.4-10) (2020-01-15)

**Note:** Version bump only for package @kongponents/kmodal





## [0.1.4-9](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-8...@kongponents/kmodal@0.1.4-9) (2019-12-16)

**Note:** Version bump only for package @kongponents/kmodal





## [0.1.4-8](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-7...@kongponents/kmodal@0.1.4-8) (2019-11-22)

**Note:** Version bump only for package @kongponents/kmodal





## [0.1.4-7](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-6...@kongponents/kmodal@0.1.4-7) (2019-11-12)

**Note:** Version bump only for package @kongponents/kmodal





## [0.1.4-6](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-5...@kongponents/kmodal@0.1.4-6) (2019-11-11)

**Note:** Version bump only for package @kongponents/kmodal





## [0.1.4-5](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-4...@kongponents/kmodal@0.1.4-5) (2019-11-08)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))





## [0.1.4-4](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-3...@kongponents/kmodal@0.1.4-4) (2019-11-08)

**Note:** Version bump only for package @kongponents/kmodal





## [0.1.4-3](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-1...@kongponents/kmodal@0.1.4-3) (2019-11-06)

**Note:** Version bump only for package @kongponents/kmodal





## [0.1.4-2](https://github.com/Kong/kongponents/compare/@kongponents/kmodal@0.1.4-1...@kongponents/kmodal@0.1.4-2) (2019-11-06)

 - [#139](https://github.com/Kong/kongponents/pull/139)
 - modal footer button margin adjusted
 - remove event listener

**Note:** Version bump only for package @kongponents/kmodal


### 0.1.4
- Update styles, KButton version 0.1.2 :arrow_right: 0.1.9
- Add props for buttons
- Add Modal docs

[#96](https://github.com/Kong/kongponents/pull/96)
