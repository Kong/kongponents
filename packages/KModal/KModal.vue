<template>
  <div
    v-if="isVisible"
    :aria-hidden="!isVisible ? 'true' : 'false'"
    class="k-modal"
    role="dialog">
    <div
      class="modal-backdrop"
      @click="close">
      <div
        class="modal-dialog"
        @click.stop>
        <div class="modal-content">
          <div class="modal-header">
            <!-- @slot Use this slot to add a Modal Header/Title -->
            <slot name="header-content">Modal Title</slot>
          </div>
          <div class="modal-body">
            <!-- @slot Use this slot to fill the body of the Modal -->
            <slot name="body-content">Modal Body</slot>
          </div>
          <div class="modal-footer">
            <!-- @slot Use this slot to place items in the footer -->
            <slot name="footer-content"/>
            <!-- @slot Use this slot to place action/proceed button -->
            <slot name="footer-actions">
              <KButton
                appearance="primary"
                @click="proceed">Proceed</KButton>
            </slot>
            <!-- @slot Use this slot to override cancel/close button -->
            <slot name="footer-dismiss">
              <KButton
                appearance="secondary"
                @click="close">Cancel</KButton>
            </slot>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import KButton from '@kongponents/kbutton'

export default {
  name: 'KModal',
  components: { KButton },

  props: {
    /**
      *  Pass whether or not the modal should be visible
      */
    isVisible: {
      type: Boolean,
      default: false
    }
  },

  mounted: function () {
    const that = this

    document.addEventListener('keydown', function (e) {
      if (that.isVisible && e.keyCode === 27) {
        that.close()
      }
    })
  },

  methods: {
    close () {
      this.$emit('close')
    },
    proceed () {
      this.$emit('proceed')
    }
  }
}
</script>

<style scoped>
  .modal-backdrop {
    position: fixed;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: rgba(0, 0, 0, 0.5);
    z-index: 1100;
  }

  .modal-dialog {
    position: relative;
    width: auto;
    max-width: 500px;
    margin: 50px auto;
    padding: 1rem;
    border-radius: 3px;
    background: #fff;
    z-index: 9999;
  }

  .modal-content {
    position: relative;
    display: flex;
    flex-direction: column;
    background-color: #fff;
    border: 0;
    border-radius: .3rem;
    outline: 0;
  }

  .modal-header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 15px 0;
    font-size: 18px;
    font-weight: 500;
    margin: 0;
    border-bottom: 1px solid #eceeef;
  }

  .modal-body {
    position: relative;
    flex: 1 1 auto;
    padding: 15px 0;
  }

  .modal-footer {
    display: flex;
    align-items: center;
    padding: 15px 0;
    border-top: 1px solid #eceeef;
  }

  .modal-footer .button:last-child {
    margin-left: 13px;
  }
</style>
