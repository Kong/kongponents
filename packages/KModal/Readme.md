```js
new Vue({
  data(){
    return {
      isVisible: false
    }
  },
  methods: {
    showModal() {
      this.isVisible = true;
    },
    closeModal() {
      this.isVisible = false;
    },
    doAThing() {
      console.log('A THING!')
    }
  },
  template:`
    <div>
      <KButton appearance='primary' :isRounded='true' :handleClick="showModal">Open Modal</KButton>
      <KModal :isVisible='isVisible' @close='closeModal'>
        <template slot="header">{{ modalHeader }}</template>
        <template slot="body">{{ modalBody }}</template>
        <template slot="footer-actions">
          <KButton appearance='danger' :isRounded='true' :handleClick='doAThing'>Do A Thing!</KButton>
        </template>
      </Kmodal>
    </div>
  `
})
```
