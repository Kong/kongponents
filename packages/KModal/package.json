{
  "name": "@kongponents/kmodal",
  "version": "0.1.3",
  "description": "Modal component",
  "main": "KModal.vue",
  "scripts": {
    "test": "jest"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "dependencies": {
    "@kongponents/kbutton": "0.1.2"
  }
}
