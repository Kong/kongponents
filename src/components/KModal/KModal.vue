<template>
  <div
    v-if="isVisible"
    :aria-label="title"
    class="k-modal"
    role="dialog"
    aria-modal="true"
  >
    <div
      class="k-modal-backdrop modal-backdrop"
      @click="(evt) => close(false, evt)"
    >
      <div class="k-modal-dialog modal-dialog">
        <div class="k-modal-content modal-content">
          <div
            v-if="$slots.title || !hideTitle"
            class="k-modal-header modal-header mb-5"
            role="heading"
            aria-level="2"
          >
            <slot name="header-content">
              {{ title }}
            </slot>
          </div>
          <div class="k-modal-body modal-body">
            <slot name="body-content">
              {{ content }}
            </slot>
          </div>
          <div class="k-modal-footer modal-footer d-flex">
            <slot name="footer-content">
              <KButton
                :appearance="cancelButtonAppearance"
                @click="close(true)"
                @keyup.esc="close(true)"
              >
                {{ cancelButtonText }}
              </KButton>
              <div class="k-modal-action-buttons">
                <slot name="action-buttons">
                  <KButton
                    :appearance="actionButtonAppearance"
                    @click="proceed"
                    @keyup.enter="proceed"
                  >
                    {{ actionButtonText }}
                  </KButton>
                </slot>
              </div>
            </slot>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, onUnmounted, watchEffect } from 'vue'
import KButton from '@/components/KButton/KButton.vue'

export default defineComponent({
  name: 'KModal',
  components: { KButton },

  props: {
    /**
     * Set the text of the title, if using title slot, this text is for the aria-label
     */
    title: {
      type: String,
      required: true,
    },
    /**
     * Title is required for aria-labelling, toggle if the title is visible on the modal
     */
    hideTitle: {
      type: Boolean,
      default: false,
    },
    /**
     * Set the text of the body content
     */
    content: {
      type: String,
      default: '',
    },
    /**
      *  Pass whether or not the modal should be visible
      */
    isVisible: {
      type: Boolean,
      default: false,
    },
    /**
     * Set the text of the action/proceed button
     */
    actionButtonText: {
      type: String,
      default: 'Submit',
    },
    /**
     * Set the appearance of the action/proceed button
     */
    actionButtonAppearance: {
      type: String,
      default: 'primary',
    },
    /**
     * Set the text of the close/cancel button
     */
    cancelButtonText: {
      type: String,
      default: 'Cancel',
    },
    /**
     * Set the appearnace of the close/cancel button
     */
    cancelButtonAppearance: {
      type: String,
      default: 'outline',
    },
  },
  emits: ['canceled', 'proceed'],

  setup(props, { emit }) {
    const handleKeydown = (e: any): void => {
      if (props.isVisible && e.keyCode === 27) {
        close(true)
      }
    }

    const close = (force = false, event?: any): void => {
      // Close if force === true or if the user clicks on .k-modal-backdrop
      if (force || event?.target?.classList?.contains('k-modal-backdrop')) {
        emit('canceled')
      }
    }

    const proceed = (): void => {
      emit('proceed')
    }

    watchEffect(() => {
      if (typeof document !== 'undefined') {
        if (props.isVisible) {
          // Hide body overflow
          document?.body?.classList.add('k-modal-overflow-hidden')
        } else {
          // Reset body overflow
          document?.body?.classList.remove('k-modal-overflow-hidden')
        }
      }
    })

    onMounted(() => {
      document.addEventListener('keydown', handleKeydown)

      if (props.isVisible) {
        // Hide body overflow
        document?.body?.classList.add('k-modal-overflow-hidden')
      }
    })

    onUnmounted(() => {
      document.removeEventListener('keydown', handleKeydown)
      // Reset body overflow
      document?.body?.classList.remove('k-modal-overflow-hidden')
    })

    return {
      close,
      proceed,
    }
  },
})
</script>

<style lang="scss" scoped>
@import '@/styles/variables';
@import '@/styles/functions';

.k-modal-backdrop {
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: var(--KModalBackdrop, rgba(11, 23, 45, .6));
  z-index: 1100;
}

// Allow modal backdrop to scroll if viewport is shorter than modal
.k-modal-overflow-hidden .k-modal-backdrop {
  overflow: auto;
}

.k-modal-dialog {
  position: relative;
  width: auto;
  max-width: var(--KModalMaxWidth, 500px);
  margin: 50px auto;
  padding: var(--spacing-xl, spacing(xl));
  border-radius: 3px;
  border: var(--KModalBorder);
  box-shadow: 0px 0px 12px 0px var(--black-10, color(black-10));
  background: #fff;
  z-index: 9999;
}

.k-modal-content {
  position: relative;
  display: flex;
  flex-direction: column;

  .k-modal-header {
    display: flex;
    align-items: center;
    justify-content: flex-start;
    color: var(--KModalHeaderColor, var(--black-500, color(black-500)));
    font-size: var(--KModalHeaderSize, 20px);
    font-weight: var(--KModalHeaderWeight, 600);
    margin-left: auto;
    margin-right: auto;
  }

  .k-modal-body {
    text-align: center;
    position: relative;
    flex: 1 1 auto;
    margin-bottom: var(--KModalBottomMargin, var(--spacing-lg, spacing(lg)));
    color: var(--KModalColor, var(--grey-500, color(grey-500)));
    font-size: var(--KModalFontSize, 13px);
    line-height: 20px;
  }

  .k-modal-footer .k-modal-action-buttons {
    margin-left: auto;
  }
}
</style>

<style lang="scss">
// Leave unscoped to target 'body' element
body.k-modal-overflow-hidden {
  overflow: hidden;
}
</style>
