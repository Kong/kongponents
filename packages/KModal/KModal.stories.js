import Vue from 'vue'
import { storiesOf } from '../../storybook/node_modules/@storybook/vue'

import KModal from './KModal.vue'

const sharedData = function () {
  return {
    isVisible: false
  }
}

const sharedMethods = {
  toggleModal (isVisible) {
    return () => {
      this.isVisible = isVisible
    }
  }
}

storiesOf('Modal', module)
  .add('Default', () => ({
    components: { KModal },
    template: `
    <div>
      <KButton @click="toggleModal(true)()">Toggle Modal</KButton>
      <KModal @close="toggleModal(false)()" :isVisible='isVisible' />
    </div>
    `,
    data: sharedData,
    methods: sharedMethods
  }), {info: {}})
  .add('Slots', () => ({
    components: { KModal },
    template: `
    <div>
      <KButton @click="toggleModal(true)()">Toggle Modal</KButton>
      <KModal @close="toggleModal(false)()" :isVisible='isVisible'>
      <template slot="header-content">
          ⚠️ Are you sure?
        </template>
        <template slot="body-content">
          This is a destructive action and by clicking this button you will destroy things.
        </template>
        <template slot="footer-actions">&#8203;</template>
        <template slot="footer-dismiss">
          <KButton style="margin-left: 0;" appearance="danger" @click="toggleModal(false)()">Destroy the things</KButton>
        </template>
      </Kmodal>
    </div>
    `,
    data: sharedData,
    methods: sharedMethods
  }), {info: {}})
