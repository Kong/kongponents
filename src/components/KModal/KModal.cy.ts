import { mount } from 'cypress/vue'
import KModal from '@/components/KModal/KModal.vue'

describe('KModal', () => {
  it('renders closed when visible is false', () => {
    mount(KModal, {
      props: {
        visible: false,
      },
    })

    cy.get('.k-modal').should('not.to.exist')
  })

  it('renders open when visible is true', () => {
    mount(KModal, {
      props: {
        visible: true,
      },
    })

    cy.get('.k-modal .modal-container').should('be.visible')
  })

  it('renders action buttons and close icon by default', () => {
    mount(KModal, {
      props: {
        visible: true,
      },
    })

    cy.getTestId('modal-close-icon').should('be.visible')
    cy.getTestId('modal-cancel-button').should('be.visible')
    cy.getTestId('modal-action-button').should('be.visible')
  })

  it('renders title when passed through prop', () => {
    const title = 'Modal Title'

    mount(KModal, {
      props: {
        visible: true,
        title,
      },
    })

    cy.get('.modal-title').should('be.visible').contains(title)
  })

  it('renders title when passed through slot', () => {
    const title = 'Slotted Title'

    mount(KModal, {
      props: {
        visible: true,
        title: 'Modal Title',
      },
      slots: {
        title,
      },
    })

    cy.get('.modal-title').should('be.visible').contains(title)
  })

  it('renders content when passed through slot', () => {
    const content = 'Modal Content'

    mount(KModal, {
      props: {
        visible: true,
      },
      slots: {
        default: content,
      },
    })

    cy.get('.modal-content').should('be.visible').contains(content)
  })

  it('renders action button properly when text, appearance and disabled props are passed', () => {
    const text = 'Action Button'
    const appearance = 'danger'
    const disabled = true

    mount(KModal, {
      props: {
        visible: true,
        actionButtonText: text,
        actionButtonAppearance: appearance,
        actionButtonDisabled: disabled,
      },
    })

    cy.getTestId('modal-action-button').should('be.visible').contains(text).should('have.class', appearance).should('be.disabled')
  })

  it('renders cancel button properly when text, appearance and disabled props are passed', () => {
    const text = 'Cancel Button'
    const appearance = 'danger'
    const disabled = true

    mount(KModal, {
      props: {
        visible: true,
        cancelButtonText: text,
        cancelButtonAppearance: appearance,
        cancelButtonDisabled: disabled,
      },
    })

    cy.getTestId('modal-cancel-button').should('be.visible').contains(text).should('have.class', appearance).should('be.disabled')
  })

  it('does not render cancel button when hideCancelButton is true', () => {
    mount(KModal, {
      props: {
        visible: true,
        hideCancelButton: true,
      },
    })

    cy.getTestId('modal-cancel-button').should('not.to.exist')
  })

  it('renders footer slot when passed', () => {
    const footer = 'Modal Footer'

    mount(KModal, {
      props: {
        visible: true,
      },
      slots: {
        footer,
      },
    })

    cy.get('.modal-footer').should('be.visible').contains(footer)
    cy.getTestId('modal-cancel-button').should('not.to.exist')
    cy.getTestId('modal-action-button').should('not.to.exist')
  })

  // footer-actions slot
  it('renders footer-actions slot when passed', () => {
    const footerActions = 'Modal Footer Actions'

    mount(KModal, {
      props: {
        visible: true,
      },
      slots: {
        'footer-actions': footerActions,
      },
    })

    cy.get('.footer-actions').should('be.visible').contains(footerActions)
    cy.getTestId('modal-cancel-button').should('not.to.exist')
    cy.getTestId('modal-action-button').should('not.to.exist')
  })

  it('renders custom content over default content when passed through modal-content slot', () => {
    const modalContent = 'Modal Content'

    mount(KModal, {
      props: {
        visible: true,
      },
      slots: {
        'modal-content': modalContent,
      },
    })

    cy.get('.modal-container').should('be.visible').contains(modalContent)
    cy.get('.modal-header').should('not.to.exist')
    cy.get('.modal-content').should('not.to.exist')
    cy.get('.modal-footer').should('not.to.exist')
  })

  it('does not render close icon when hideCloseIcon is true', () => {
    mount(KModal, {
      props: {
        visible: true,
        hideCloseIcon: true,
      },
    })

    cy.getTestId('modal-close-icon').should('not.to.exist')
  })

  it('renders modal with correct width when prop is passed', () => {
    const width = '123px'

    mount(KModal, {
      props: {
        visible: true,
        width,
      },
    })

    cy.get('.k-modal .modal-container').should('have.css', 'width', width)
  })

  it('renders modal with correct max-height when prop is passed', () => {
    const maxHeight = '123px'

    mount(KModal, {
      props: {
        visible: true,
        maxHeight,
      },
    })

    cy.get('.k-modal .modal-content').should('have.css', 'max-height', maxHeight)
  })

  it('emits proceed event when action button is clicked', () => {
    mount(KModal, {
      props: {
        visible: true,
      },
    })

    cy.getTestId('modal-action-button').click().then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'proceed')
    })
  })

  it('emits canceled event when cancel button is clicked', () => {
    mount(KModal, {
      props: {
        visible: true,
      },
    })

    cy.getTestId('modal-cancel-button').click().then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'canceled')
    })
  })

  it('does not close modal on backdrop click when closeOnBackdropClick is false', () => {
    mount(KModal, {
      props: {
        visible: true,
        closeOnBackdropClick: false,
      },
    })

    cy.get('.k-modal .modal-backdrop').click('topRight').then(() => {
      cy.get('.k-modal .modal-container').should('be.visible')
    })
  })

  it('emits canceled event when backdrop is clicked and closeOnBackdropClick is true', () => {
    mount(KModal, {
      props: {
        visible: true,
        closeOnBackdropClick: true,
      },
    })

    cy.get('.k-modal .modal-backdrop').click('topRight').then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'canceled')
    })
  })

  it('automatically focuses in the first input element on modal open when present', () => {
    mount(KModal, {
      props: {
        visible: true,
      },
      slots: {
        default: '<input type="text" data-testid="first-input" /><input type="text" data-testid="second-input" />',
      },
    })

    cy.getTestId('first-input').should('be.focused')
  })
})
