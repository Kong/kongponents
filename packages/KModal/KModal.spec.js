import { mount } from '@vue/test-utils'
import KModal from '@/KModal/KModal'

describe('KModal', () => {
  it('renders header slot when passed', () => {
    const headerText = 'This is some header text'
    const wrapper = mount(KModal, {
      propsData: {
        isVisible: true
      },
      slots: {
        'header-content': `<div>${headerText}</div>`
      }
    })

    expect(wrapper.find('.modal-header').html())
      .toEqual(expect.stringContaining(headerText))
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('matches snapshot', () => {
    const wrapper = mount(KModal, {
      propsData: {
        isVisible: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
