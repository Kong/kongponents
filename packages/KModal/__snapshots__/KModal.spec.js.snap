// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KModal matches snapshot 1`] = `
<div aria-hidden="false" role="dialog" class="k-modal">
  <div class="modal-backdrop">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">Modal Title</div>
        <div class="modal-body">Modal Body</div>
        <div class="modal-footer"> <button class="button rounded primary">Proceed</button> <button class="button rounded secondary">Cancel</button></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`KModal renders header slot when passed 1`] = `
<div aria-hidden="false" role="dialog" class="k-modal">
  <div class="modal-backdrop">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <div>This is some header text</div>
        </div>
        <div class="modal-body">Modal Body</div>
        <div class="modal-footer"> <button class="button rounded primary">Proceed</button> <button class="button rounded secondary">Cancel</button></div>
      </div>
    </div>
  </div>
</div>
`;
